### ISSUE TITLE
norm of a vector should use `inner_product_matrix`

### ISSUE THREAD

**Issue by jhpalmieri** [sagemath/sage#35400]
### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### Problem Description

The `norm` method for vectors ignores the value of `inner_product_matrix`:

```
sage: B = Matrix(QQ,3,3,[1,0,0, 0,2,0, 0,0,3])
sage: V3 = VectorSpace(QQ, 3, inner_product_matrix = B)
sage: v = V3((1,2,3))
sage: v.norm()
sqrt(14)
sage: sqrt(v.inner_product(v))
6
```

### Proposed Solution

The behavior should be changed: `v.norm()` should return the same as `sqrt(v.inner_product(v))`.

### Alternatives Considered

Alternatively the documentation should make this behavior clear, explicitly pointing out that `inner_product_matrix` is ignored, with examples.

**Comment by karan0handa** [comment]
Some possible solutions that come to my mind - 
1. Add a boolean parameter `use_inner_product_matrix=True` to `norm` and return the `sqrt(inner_product(self))`
2. Rename the current method `pnorm` and create a new method `norm` that is purely based on the inner product matrix.

First one sounds more reasonable to me before there may be a lot of additional dependencies to handle with the second solution. Let me know what you think @jhpalmieri @mkoeppe.

**Comment by jhpalmieri** [comment]
I like the idea of adding a boolean parameter to the current method. To maintain backward-compatibility, it might be good to start with `use_inner_product_matrix=False` with a deprecation period, after which it switches to `True`. (Someone may be relying on the current behavior in their own code.)

**Comment by mkoeppe** [comment]
Not clear to me that this by itself is a good deprecation path. Maybe for the quadratic modules, make `use_inner_product_matrix` a required keyword argument and issue a deprecation warning if it is not passed?

**Comment by mkoeppe** [comment]
Or instead of a boolean, allow users to pass an `inner_product_matrix` to `norm`, with special values ``False`` (= use Euclidean), ``True`` (= use the inner product matrix of the module), and deprecation warning if not passed.

**Comment by mkoeppe** [comment]
The interface is complicated by the fact that the method `norm` also accepts the following positional parameter:
```
   * "p" - default: 2 - "p" can be a real number greater than 1,
     infinity ("oo" or "Infinity"), or a symbolic expression.
```

**Comment by karan0handa** [comment]
> The interface is complicated by the fact that the method `norm` also accepts the following positional parameter:
> 
> ```
>    * "p" - default: 2 - "p" can be a real number greater than 1,
>      infinity ("oo" or "Infinity"), or a symbolic expression.
> ```

Yeah, the positional parameter makes the current `norm` definition very restricted because of the type of the parameter being a real number, however the same parameter can also take string types as is done in [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.norm.html#scipy.linalg.norm) and [numpy](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html).

In short, we change the argument from `p` to `ord` (or some other name) which accepts both string and int values and one of the values is `use_inner_product_matrix` which uses the module's inner_product_matrix along with appropriate deprecation warnings.

**Comment by karan0handa** [comment]
@mkoeppe Let me know if the above solution works and I'll create a PR.

**Comment by mkoeppe** [comment]
I would prefer a separate keyword parameter `inner_product_matrix` as I suggested in https://github.com/sagemath/sage/issues/35400#issuecomment-1498428090
